### PR TITLE
Implement type safe configuration-as-code integration

### PR DESCRIPTION
**Pull request by timja**
Adds type safe config rather than strings with separators, example:

```
jenkins:
  authorizationStrategy:
    globalMatrix:
      permissions:
      - permission: "Overall/Administer"
        user: "admin1"
      - permission: "Overall/Administer"
        user: "admin"
      - group: "my-group"
        permission: "Overall/Read"
```

No compatibility added for previous format.
Drops support for `grantedPermissions`

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/matrixauth/integrations/casc/MatrixAuthorizationStrategyConfigurator.java [hunk 1]
```diff
@@ -29,47 +34,43 @@
...
+                        .map(v -> {
+                            PermissionEntryForCasc entry = new PermissionEntryForCasc(e.getKey().group.getId() + "/" + e.getKey().name);
+                            if (v.getType().equals(AuthorizationType.USER)) {
+                                entry.setUser(v.getSid());
+                                return entry;
+                            } else if (v.getType().equals(AuthorizationType.GROUP)) {
+                                entry.setGroup(v.getSid());
+                                return entry;
+                            }
+                            throw new IllegalStateException("Ambiguous users are not supported: " + v.getSid());
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/matrixauth/integrations/casc/MatrixAuthorizationStrategyConfigurator.java:62]: Causes a failure to export if there's any `either` types
image

Could:
- Add support for either with a log message maybe?
- Filter either out
- or just have it error like this? not very nice UX though, thoughts?

**Review by timja — COMMENTED** [review]
@daniel-beck any thoughts on this?

example config file:

```yaml
jenkins:
  authorizationStrategy:
    globalMatrix:
      permissions:
      - permission: "Overall/Administer"
        user: "admin1"
      - permission: "Overall/Administer"
        user: "admin"
      - group: "my-group"
        permission: "Overall/Read"
```

**Comment by timja** [comment]
@daniel-beck any thoughts?

**Comment by daniel-beck** [comment]
@timja Thanks for the ping, I missed the original as I was offline for a while then. Looks nice, thanks. I'll review it further and perhaps look into amending the job-dsl syntax in a similar manner.

**Comment by timja** [comment]
@daniel-beck any interest in this?

**Comment by daniel-beck** [comment]
@timja Definitely, just need to find the time :)

**Comment by timja** [comment]
@daniel-beck this is getting close to it's birthday 😢, any chance of a review or would you approve an adoption pull request?

**Comment by daniel-beck** [comment]
> any chance of a review

Still on my list. Keeps getting pre-empted by other, more urgent work, unfortunately.

> would you approve an adoption pull request?

Definitely not a single-purpose adoption.

**Comment by daniel-beck** [comment]
> Noting that there should be a corresponding change to the Job DSL support so it looks vaguely similar. Unsure so far how to accomplish.

Some progress here:

Job DSL can look like the following fairly easily:

```
folder('generated') {
  properties {
    authorizationMatrix {
      inheritanceStrategy {
          nonInheriting()
      }
      entries {
        group {
          name('authenticated')
          permissions([
            'Job/Build',
            'Job/Configure',
            'Job/Read'
          ])}
        user {
          name('admin')
          permissions(['Job/Delete'])
        }
      }
    }
  }
}
```

The corresponding Pipeline snippet looks like this, according to Snippet generator (formatted for readability):

```
properties([
  authorizationMatrix([
    user(name: 'admin', permissions: ['hudson.model.Item.Delete']),
    group(name: 'authenticated', permissions: [
      'hudson.model.Item.Read',
      'hudson.model.Item.Build',
      'hudson.model.Item.Configure'
    ])
  ])
])

----

I'm kinda worried about the symbols here though because of `configuration-as-code` incorrectly implementing the symbol scopes (hence `mavenGlobalConfig`). Is that still the case @timja ?

**Comment by daniel-beck** [comment]
Closing in favor of #145 which seems mostly done.

**Pull request closed without merging**